lib/permessage-deflate.js

```diff
@@ -412,12 +411,20 @@ class PerMessageDeflate {
         ? zlib.Z_DEFAULT_WINDOWBITS
         : this.params[key];
 
-      this._deflate = zlib.createDeflateRaw({
+      const zlibOptions = {
+        // TODO deprecate memLevel/level and recommend zlibOptions instead
         memLevel: this._options.memLevel,
         level: this._options.level,
-        flush: zlib.Z_SYNC_FLUSH,
         windowBits
-      });
+      };
+      if (this._options.zlibOptions != null) {
+        if (typeof this._options.zlibOptions === 'number') {
+          throw new Error('Do not set `windowBits` on `zlibOptions`!');
+        }
+        Object.assign(zlibOptions, this._options.zlibOptions);
+      }
+
+      this._deflate = zlib.createDeflateRaw(zlibOptions);
```

Comment:
I would just pass the options object here, overriding the value of the `windowBits`.

```js
this._deflate = zlib.createDeflateRaw(
  Object.assign(
    {
      memLevel: this._options.memLevel,
      level: this._options.level,
      flush: zlib.Z_SYNC_FLUSH
    },
    this._options.zlibOption,
    { windowBits }
  )
);
```

Reply:
The idea was to avoid the Object.assign() if nothing is passed (the far more common case).

Reply:
Is the perf impact big? The next version will probably be a major due to https://github.com/websockets/ws/pull/1302 so backward compatibility for `options.memLevel` and `options.level` can be removed.

Reply:
Actually no, it won't be, because it's only per-connection, I had erroneously assumed it was per message. Will update.

Reply:
Updated; one thing to note, setting `flush` should not be necessary because we pass `zlib.Z_SYNC_FLUSH` when flushing manually; please correct me if wrong.

Reply:
I was about to ask exactly this :smile: . I think you are correct bit I'm not 100% sure.

Reply:
You are also correct wrt to the stream instance. It is instantiated per message if context takeover is disabled.